From a44c1117d89958dd72123d07440281ec91f6f36a Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Tue, 19 Aug 2025 10:17:48 +0200
Subject: a11y iOS: Announce changes to name/description to VoiceOver

Unlike macOS AppKit iOS UIKit does not offer an API to notify
VoiceOver about a value/title change. Also there is no public
API to query the system for the element which is currently
focused.

Both these caveats are addressed in this commit: the element with
active a11y focus is tracked; if the Accessible.name or
Accessible.description for *this* element changes, VoiceOver
is informed via UIAccessibilityLayoutChangedNotification and
refocuses the updated element.

Task-number: QTBUG-139275
Pick-to: 6.8 6.9 6.10
Change-Id: Ief1a2928aac138fe3f7afd9571838ed7713b327e
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 6d0c380ac12c9f87fba464539a788b3185e85fb9)
---
 .../platforms/ios/qiosplatformaccessibility.h     |  6 ++++++
 .../platforms/ios/qiosplatformaccessibility.mm    | 15 ++++++++++++++-
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git x/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.h y/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.h
index 96efc663bac896adb44a80eb5953a57ec7d92381..49cecbfec1f19a33435293f3a389e630c46fcae3 100644
--- x/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.h
+++ y/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.h
@@ -5,6 +5,8 @@
 #define QIOSPLATFORMACCESSIBILITY_H
 
 #include <qpa/qplatformaccessibility.h>
+#include <QtCore/private/qcore_mac_p.h>
+#include <quiaccessibilityelement.h>
 
 #if QT_CONFIG(accessibility)
 
@@ -17,6 +19,10 @@ public:
     ~QIOSPlatformAccessibility();
 
     virtual void notifyAccessibilityUpdate(QAccessibleEvent *event);
+
+private:
+    QMacNotificationObserver m_focusObserver;
+    QMacAccessibilityElement *m_focusElement;
 };
 
 QT_END_NAMESPACE
diff --git x/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.mm y/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.mm
index f661db7a2e19a35dac076970641bce4a72b564b4..6869db5ae81411b42be9d293327acab3a39faec2 100644
--- x/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.mm
+++ y/qtbase/src/plugins/platforms/ios/qiosplatformaccessibility.mm
@@ -12,7 +12,13 @@
 #include "quiaccessibilityelement.h"
 
 QIOSPlatformAccessibility::QIOSPlatformAccessibility()
-{}
+{
+    m_focusObserver = QMacNotificationObserver(
+            nil, UIAccessibilityElementFocusedNotification, [&](NSNotification *notification) {
+                id element = notification.userInfo[UIAccessibilityFocusedElementKey];
+                m_focusElement = static_cast<QMacAccessibilityElement *>(element);
+            });
+}
 
 QIOSPlatformAccessibility::~QIOSPlatformAccessibility()
 {}
@@ -59,6 +65,13 @@ void QIOSPlatformAccessibility::notifyAccessibilityUpdate(QAccessibleEvent *even
         UIAccessibilityPostNotification(UIAccessibilityLayoutChangedNotification, element);
         break;
     }
+    case QAccessible::DescriptionChanged:
+    case QAccessible::NameChanged: {
+        auto *element = [QMacAccessibilityElement elementWithId:event->uniqueId()];
+        if (element == m_focusElement)
+            UIAccessibilityPostNotification(UIAccessibilityLayoutChangedNotification, element);
+        break;
+    }
     case QAccessible::ObjectCreated:
     case QAccessible::ObjectShow:
     case QAccessible::ObjectHide:
